### PR TITLE
MCR-3092 XSL test capabilities

### PR DESCRIPTION
**Pull request by michael-becker**
[Link to jira](https://mycore.atlassian.net/browse/MCR-3092).

MCRTestCase contains new methods
* parse: parses a given XML structure using a given XSL file
* xmlTest: encapsulates an XML structure in an XML document for test

Usage:
* create a test.xsl file containing match templates calling functions/templates to test (e.g. classificationTest.xsl)
* parse a structure/empty XML using the parse-method in MCRTestCase
* call Junit-assert-statements using the resulting XML structure

see usage example in XSLClassificationTest

### CODE REVIEW

File: mycore-base/src/test/resources/xslt/functions/classificationTest.xsl [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+
+  <xsl:template match="test-current-label-text">
+    <xsl:variable name="class" select="mcrclass:category($classid, $categid)"/>
+    <result>
+      <xsl:value-of select="mcrclass:current-label-text($class)"/>
+    </result>
+  </xsl:template>
+
+</xsl:stylesheet>
```
Comment (rsteph-de) [inline mycore-base/src/test/resources/xslt/functions/classificationTest.xsl]: please add newline at end of file

File: mycore-base/src/test/java/org/mycore/common/MCRTestCase.java [hunk 2]
```diff
@@ -200,4 +211,45 @@ private String buildFileLocation(String fileName) {
...
+        root.addContent(xml.detach());
+
+        return new Document(root);
+    }
+
+    /**
+     * Parses the XML document using the given XSL stylesheet with the given parameters.
+     *
+     * @param parameters the XSL parameters
+     * @param xsl        the XSL stylesheet file used for parsing
```
Comment (rsteph-de) [inline mycore-base/src/test/java/org/mycore/common/MCRTestCase.java]: add "from classpath" or "as resource path" to be clear where XSL file comes from

File: mycore-base/src/test/java/org/mycore/common/MCRTestCase.java [hunk 3]
```diff
@@ -200,4 +211,45 @@ private String buildFileLocation(String fileName) {
...
+     * @param xml        the XML document to parse
+     */
+    protected Document parse(Map<String, Object> parameters, String xsl, Document xml) throws TransformerException {
```
Comment (rsteph-de) [inline mycore-base/src/test/java/org/mycore/common/MCRTestCase.java]: maybe rename to transform()

File: mycore-base/src/test/java/org/mycore/common/MCRJPATestCase.java [hunk 4]
```diff
@@ -249,4 +254,14 @@ protected static void executeUpdate(String sql, Consumer<Integer> consumer) {
         });
     }
 
+    /**
+     * Adds a classification represented by the given XML file to the system.
+     *
+     * @param file the XML file containing the classification
```
Comment (rsteph-de) [inline mycore-base/src/test/java/org/mycore/common/MCRJPATestCase.java]: add "from classpath"

File: mycore-base/src/test/java/org/mycore/common/MCRJPATestCase.java [hunk 5]
```diff
@@ -249,4 +254,14 @@ protected static void executeUpdate(String sql, Consumer<Integer> consumer) {
...
+     */
+    protected void addClassification(String file) throws Exception {
```
Comment (rsteph-de) [inline mycore-base/src/test/java/org/mycore/common/MCRJPATestCase.java]: rename "file" maybe to resourcePath

File: mycore-base/src/test/java/org/mycore/common/MCRJPATestCase.java [hunk 6]
```diff
@@ -249,4 +254,14 @@ protected static void executeUpdate(String sql, Consumer<Integer> consumer) {
...
+        Document classification = new SAXBuilder().build(getClass().getResourceAsStream(file));
+        MCRCategory category = MCRXMLTransformer.getCategory(classification);
+        MCRCategoryDAOFactory.getInstance().addCategory(null, category);
+    }
```
Comment (yagee-de) [inline mycore-base/src/test/java/org/mycore/common/MCRJPATestCase.java]: Should not be part of MCRJPATestCase as Classifications has nothing to do with providing JPA functionality

File: mycore-base/src/test/java/org/mycore/common/MCRTestCase.java [hunk 7]
```diff
@@ -200,4 +211,45 @@ private String buildFileLocation(String fileName) {
...
+
+        if (transformer instanceof TransformerImpl transformerImpl) {
+            transformerImpl.getUnderlyingXsltTransformer()
+                .setMessageHandler(msg -> System.err.println(msg.getContent().getStringValue()));
+        }
+
+        transformer.transform(new JDOMSource(xml), result);
+
+        return new Document(result.getResult());
+    }
```
Comment (yagee-de) [inline mycore-base/src/test/java/org/mycore/common/MCRTestCase.java]: Should not be method in an MCRTestCase. Maybe you want to provide an XSLTHelper class in mycore-base/src/test/java instead?

File: mycore-base/src/test/java/org/mycore/common/MCRTestCase.java (hunk 3 above)
Comment (yagee-de) [inline mycore-base/src/test/java/org/mycore/common/MCRTestCase.java]: It does not parse the XML but it transforms it. So a better name would be transform. The parameter order could also be improved: transform XML with XSL using these PARAMETERS

File: mycore-base/src/test/java/org/mycore/frontend/xsl/XSLClassificationTest.java [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+        result = parse(Map.of("classid", "TestClassification", "categid", "junit_en"), XSL,
+            xmlTest("test-current-label-text", empty)).getRootElement();
+        assertEquals("junit_en (en)", result.getText());
+
+        // we neither have current lang nor default lang label
+        result = parse(Map.of("classid", "TestClassification", "categid", "junit_none"), XSL,
+            xmlTest("test-current-label-text", empty)).getRootElement();
+        assertEquals("??junit_none@de??", result.getText());
+    }
+}
```
Comment (yagee-de) [inline mycore-base/src/test/java/org/mycore/frontend/xsl/XSLClassificationTest.java]: This is really hard to read and to understand. Did you consider using xspec?

**Review by yagee-de — CHANGES_REQUESTED** [review]

File: mycore-base/src/test/java/org/mycore/common/MCRJPATestCase.java (hunk 6 above)
Comment (michael-becker) [reply mycore-base/src/test/java/org/mycore/common/MCRJPATestCase.java]: the method is now in an util class (MCRTestCaseClassificationUtil)

File: mycore-base/src/test/java/org/mycore/common/MCRTestCase.java (hunk 7 above)
Comment (michael-becker) [reply mycore-base/src/test/java/org/mycore/common/MCRTestCase.java]: both methods are now in an util class (MCRTestCaseXSLTUtil)

File: mycore-base/src/test/java/org/mycore/frontend/xsl/XSLClassificationTest.java (hunk 8 above)
Comment (michael-becker) [reply mycore-base/src/test/java/org/mycore/frontend/xsl/XSLClassificationTest.java]: If I get this right, it is not possible to simly integrate xspec in junit test cases. The advantage of this approach is that we do not have antother external dependency and we can run tests without calling an external application.

File: mycore-base/src/test/java/org/mycore/common/util/MCRTestCaseXSLTUtil.java [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+
+        TransformerFactory factory = TransformerFactory.newInstance();
+        factory.setURIResolver(MCRURIResolver.instance());
+
+        Transformer transformer = factory.newTransformer(xslt);
+        parameters.forEach(transformer::setParameter);
+
+        if (transformer instanceof TransformerImpl transformerImpl) {
+            transformerImpl.getUnderlyingXsltTransformer()
+                .setMessageHandler(msg -> System.err.println(msg.getContent().getStringValue()));
```
Comment (rsteph-de) [inline mycore-base/src/test/java/org/mycore/common/util/MCRTestCaseXSLTUtil.java]: use LOGGER.error()

File: mycore-base/src/test/java/org/mycore/common/util/MCRTestCaseXSLTUtil.java [hunk 10]
```diff
@@ -32,28 +36,50 @@
...
-     * @param method the method name used as root element name
-     * @param xml    the XML tree
-     * @return an XML document with the method name as root element and the content of the given XML tree as children
+     * @param rootName the name used as root element name
+     * @param xml      the XML tree
+     * @return an XML document with the name as root element and the content of the given XML tree as children
      */
-    public static Document xmlTest(String method, Element xml) {
-        final Element root = new Element(method);
+    public static Document prepareTestDocument(String rootName, Element xml) {
```
Comment (rsteph-de) [inline mycore-base/src/test/java/org/mycore/common/util/MCRTestCaseXSLTUtil.java:73]: Idea: to be more flexible: the 2nd param could be a Content list (TextNodes or Elements)
or provide another prepareTestDocument() method ?

**Review by rsteph-de — CHANGES_REQUESTED** [review]
Please add missing Licence Header in Java files

File: mycore-base/src/test/java/org/mycore/frontend/xsl/XSLClassificationTest.java [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+import org.jdom2.Document;
+import org.jdom2.Element;
+import org.junit.Test;
+import org.mycore.common.MCRJPATestCase;
+
+import java.util.Map;
+
+import static org.junit.Assert.assertEquals;
+
+public class XSLClassificationTest extends MCRJPATestCase {
```
Comment (rsteph-de) [inline mycore-base/src/test/java/org/mycore/frontend/xsl/XSLClassificationTest.java]: rename to MCRXSLClassificationTest

File: mycore-base/src/test/resources/xslt/functions/classificationTest.xsl [hunk 12]
```diff
@@ -0,0 +1,22 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<xsl:stylesheet version="2.0"
```
Comment (rsteph-de) [inline mycore-base/src/test/resources/xslt/functions/classificationTest.xsl]: use version= "3.0" - we have never used a 2.0 complient XSLT processor

**Review by rsteph-de — CHANGES_REQUESTED** [review]

**Review by kkrebs — APPROVED** [review]

**Review by yagee-de — APPROVED** [review]

**Review by rsteph-de — APPROVED** [review]
